File: backend/routers/companies.js
const express = require('express');
const fs = require('fs');
const path = require('path');

const router = express.Router();
const DATA_PATH = path.join(__dirname, '..', 'data', 'companies_data.json');

function loadCompanies() {
  if (!fs.existsSync(DATA_PATH)) {
    return [];
  }
  try {
    return JSON.parse(fs.readFileSync(DATA_PATH, 'utf-8'));
  } catch (err) {
    console.error('Error reading companies data:', err);
    return [];
  }
}

router.get('/', (req, res) => {
  const sector = req.query.sector || 'All';
  let companies = loadCompanies();

  if (sector && sector !== 'All') {
    companies = companies.filter(
      c => (c.sector || '').toLowerCase() === sector.toLowerCase()
    );
  }

  // Return summarized company cards
  const summaries = companies.map(c => {
    const crisis = c.crisis_info || {};
    const financials = c.six_year_financials || [{}];
    const latestFin = financials[financials.length - 1] || {};
    const moves = c.six_year_downfalls_and_pumps || [];
    const downfalls = moves.filter(m => m.type === 'Downfall');
    const pumps = moves.filter(m => m.type === 'Pump' || m.type === 'Rebound');

    return {
      id: c.id,
      name: c.name,
      ticker: c.ticker,
      sector: c.sector,
      logo_color: c.logo_color,
      logo_url: c.logo_url,
      market_cap: c.market_cap,
      founded_year: c.founded_year,
      founder: c.leadership?.founder,
      current_ceo: c.leadership?.current_ceo,
      executive_analysis_paragraph: c.executive_analysis_paragraph,
      downfalls_count: downfalls.length,
      pumps_count: pumps.length,
      primary_downfall: downfalls[0]?.title || null,
      primary_pump: pumps[0]?.title || null,
      six_year_downfalls_and_pumps: moves,
      crisis_title: crisis.crisis_title,
      strategy_used: crisis.strategy_used,
      recovery_status: crisis.recovery_status,
      financial_summary: {
        latest_revenue_growth: latestFin.revenue_growth ?? 0,
        latest_roe: latestFin.roe ?? 0,
        latest_cash_flow: latestFin.cash_flow ?? 0,
        start_year: financials[0]?.year || '2019',
        end_year: latestFin.year || '2024'
      }
    };
  });

  res.json({
    sector_filter: sector,
    total_companies: summaries.length,
    companies: summaries
  });
});

router.get('/sectors', (req, res) => {
  res.json([
    'All',
    'Technology',
    'Retail',
    'FinTech',
    'Real Estate',
    'Automobile',
    'Healthcare',
    'Media'
  ]);
});

router.get('/:companyId', (req, res) => {
  const { companyId } = req.params;
  const companies = loadCompanies();
  const company = companies.find(c => c.id === companyId);

  if (!company) {
    return res.status(404).json({ detail: `Company '${companyId}' not found` });
  }

  res.json(company);
});

module.exports = router;

File: backend/routers/companies.py
import json
import os
from fastapi import APIRouter, Query, HTTPException
from typing import Optional, List

router = APIRouter(prefix="/api/companies", tags=["Company Intelligence"])

DATA_PATH = os.path.join(os.path.dirname(__file__), "..", "data", "companies_data.json")

def load_companies():
    if not os.path.exists(DATA_PATH):
        return []
    with open(DATA_PATH, "r", encoding="utf-8") as f:
        return json.load(f)

@router.get("/")
def get_all_companies(sector: Optional[str] = Query("All", description="Sector filter")):
    companies = load_companies()
    if sector and sector != "All":
        companies = [c for c in companies if c.get("sector", "").lower() == sector.lower()]
    
    # Return summarized company cards
    summaries = []
    for c in companies:
        crisis = c.get("crisis_info", {})
        summaries.append({
            "id": c.get("id"),
            "name": c.get("name"),
            "ticker": c.get("ticker"),
            "sector": c.get("sector"),
            "logo_color": c.get("logo_color"),
            "market_cap": c.get("market_cap"),
            "founded_year": c.get("founded_year"),
            "founder": c.get("leadership", {}).get("founder"),
            "current_ceo": c.get("leadership", {}).get("current_ceo"),
            "crisis_title": crisis.get("crisis_title"),
            "strategy_used": crisis.get("strategy_used"),
            "recovery_status": crisis.get("recovery_status"),
            "financial_summary": {
                "latest_revenue_growth": c.get("six_year_financials", [{}])[-1].get("revenue_growth", 0),
                "latest_roe": c.get("six_year_financials", [{}])[-1].get("roe", 0),
                "latest_cash_flow": c.get("six_year_financials", [{}])[-1].get("cash_flow", 0)
            }
        })
    
    return {
        "sector_filter": sector,
        "total_companies": len(summaries),
        "companies": summaries
    }

@router.get("/sectors")
def get_available_sectors():
    return [
        "All",
        "Technology",
        "Retail",
        "FinTech",
        "Real Estate",
        "Automobile",
        "Healthcare",
        "Media"
    ]

@router.get("/{company_id}")
def get_company_detail(company_id: str):
    companies = load_companies()
    for c in companies:
        if c.get("id") == company_id:
            return c
    raise HTTPException(status_code=404, detail=f"Company '{company_id}' not found")
